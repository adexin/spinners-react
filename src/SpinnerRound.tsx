import React, { CSSProperties } from 'react';
import { defaultProps, SpinnerProps } from './helpers';
import { SpinnersProps, withSharedProps } from './withSharedProps';

import './SpinnerRound.css';

export type SpinnerRoundProps = SpinnersProps & SpinnerProps;

const Component = ({
  speed = defaultProps.speed,
  still = defaultProps.still,
  thickness = defaultProps.thickness,
  ...svgProps
}: SpinnerProps) => {
  const strokeWidth = 3 * (thickness / 100);
  const circleStyle: CSSProperties = {
    animation: `spinners-react-round ${140 / speed}s ease-in-out infinite`,
    transformOrigin: 'center',
  };

  if (still) circleStyle.animation = 'none';

  return (
    <svg fill="none" {...svgProps} viewBox="0 0 66 66">
      <circle
        cx="33"
        cy="33"
        fill="none"
        r={28}
        stroke="currentColor"
        strokeWidth={strokeWidth}
        style={circleStyle}
      />
    </svg>
  );
};

export const SpinnerRound = withSharedProps(Component);
