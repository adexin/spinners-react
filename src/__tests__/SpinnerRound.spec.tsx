import { render } from '@testing-library/react';
import React from 'react';

import { SpinnerRound } from '../SpinnerRound';

describe('SpinnerRound', () => {
  const color = 'red';

  it('matches the snapshot', () => {
    const { container } = render(
      <SpinnerRound color="#fff" size={100} speed={10} thickness={50} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot with default props', () => {
    const { container } = render(<SpinnerRound />);

    expect(container).toMatchSnapshot();
  });

  it('renders null if disabled', () => {
    const { container } = render(<SpinnerRound enabled={false} />);

    expect(container.firstChild).toBe(null);
  });

  it('renders still if specified', () => {
    const { container } = render(<SpinnerRound still />);
    const circle = container.querySelector('circle');

    expect(circle?.style.animation).toBe('none');
  });

  it('passes props to styles', () => {
    const size = '20%';
    const thickness = 40;
    const speed = 50;
    const { container } = render(
      <SpinnerRound
        color={color}
        size={size}
        speed={speed}
        thickness={thickness}
      />,
    );
    const circle = container.querySelector('circle');

    expect(container.firstChild).toHaveStyle({ width: size });
    expect(circle).toHaveAttribute('stroke-width', `${3 * (thickness / 100)}`);
    expect(circle?.style.animation.includes(`${140 / speed}`)).toBe(true);
  });

  it('passes svg props overriding styles', () => {
    const className = 'test-class';
    const { container } = render(
      <SpinnerRound className={className} color="green" style={{ color }} />,
    );

    expect(container.firstChild).toHaveClass(className);
    expect(container.firstChild).toHaveStyle({ color });
  });
});
