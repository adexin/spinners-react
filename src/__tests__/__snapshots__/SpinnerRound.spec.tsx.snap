// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerRound matches the snapshot 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(255, 255, 255); overflow: visible; width: 100px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="currentColor"
      stroke-width="1.5"
      style="animation: spinners-react-round 14s ease-in-out infinite; transform-origin: center;"
    />
  </svg>
</div>
`;

exports[`SpinnerRound matches the snapshot with default props 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(56, 173, 72); overflow: visible; width: 50px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="currentColor"
      stroke-width="3"
      style="animation: spinners-react-round 1.4s ease-in-out infinite; transform-origin: center;"
    />
  </svg>
</div>
`;
